File: 0041_first_missing_positive.py
import heapq

class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        heapq.heapify(nums)
        start = 1
        while nums:
            pop = heapq.heappop(nums)
            if pop < start:
                continue
            elif pop > start:
                return start
            else: 
                start += 1
        if pop < 0: 
            return 1
        else: 
            return pop + 1

File: 0001-0500/0041_firstMissingPositive.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        int i = 0;
        int temp;

        while (i < nums.length) {
            while (0 < nums[i] && nums[i] <= nums.length && i + 1 != nums[i] && nums[i] != nums[nums[i] - 1]) {
                temp = nums[i];
                nums[i] = nums[temp - 1];
                nums[temp - 1] = temp;
            }
            i++;
        }

        for (int j = 0; j < nums.length; j++) {
            if (nums[j] != j + 1) {
                return j + 1;
            }
        }

        return nums.length + 1;
    }
}